src/main/java/edu/harvard/iq/dataverse/engine/command/impl/ReconcileDatasetPidCommand.java

```diff
@@ -0,0 +1,181 @@
+        if (!newPidProvider.registerWhenPublished()) {
+            registerExternalIdentifier(getDataset(), ctxt, true); // this updates GlobalIdCreateTime and IdentifierRegistered
+        }
+        // keep old Id als alternative identifier
+        AlternativePersistentIdentifier api;
+        api = new AlternativePersistentIdentifier();
+        api.setProtocol(oldId.getProtocol());
+        api.setAuthority(oldId.getAuthority());
+        api.setIdentifier(oldId.getIdentifier());
+        api.setDvObject(getDataset());
+        api.setStorageLocationDesignator(true);// cf. Dataset#getIdentifierForFileStorage()
+        if(getDataset().getAlternativePersistentIndentifiers()!=null) {
+            getDataset().getAlternativePersistentIndentifiers().add(api);
+        }else{
+            getDataset().setAlternativePersistentIndentifiers(Set.of(api));
+        }
+        // We keep the old persistent identifier as AlternativePersistentIdentifier with storageLocationDesignator true.
+        // This keep the link the object store intact, without altering the files.
+        // IMHO: This command should also update the storage. First, maintenance of the storage becomes a mess if there is a counterintuitive layout.
+        // Second, it could occur that another object is minted with the old identifier and consequently we have a conflict in our storage system.
+        // We accept this risk for now since a super user can update the storage manually from old file path to the new one, and remove the AlternativeIdentifer form the database.
+        // This removed the old identifier totally from the system and avoids all side effects...
+        // @TODO: Move files on storage once the storage API support a move operation!
+        ctxt.em().merge(getDataset());
+        ctxt.em().flush();
+
+        logger.fine("Reconciling dataset( id =`" + getDataset().getId() + ")` - Replaced old globalId `" + oldId + " with new globalId `" + getDataset().getGlobalId() + '`');
+        // notify all users with direct role assignments about the changed persistent identifier
+        List<RoleAssignment> ras = ctxt.roles().directRoleAssignments(getDataset());
+        for (RoleAssignment ra : ras) {
+            for (AuthenticatedUser au : ctxt.roleAssignees().getExplicitUsers(ctxt.roleAssignees().getRoleAssignee(ra.getAssigneeIdentifier()))) {
+                ctxt.notifications().sendNotification(au, new Timestamp(new Date().getTime()), UserNotification.Type.PIDRECONCILED,getDataset().getLatestVersion().getId(),"Persistent identifier changed!");
+            }
+        }
+
+        return getDataset();
+    }
+
+    private void reconcileFilePids(CommandContext ctxt) {
+        // remove datafile PIDs
+            try {
+                for (DataFile df : getDataset().getFiles()) {
+                    GlobalId oldPid=df.getGlobalId();
```

Comment:
Files may not have PIDs and, pre-#7334, shouldn't have PIDs before publication. Minimally need to have a null check if you want to keep the code ready for when files get registered PIDs at upload.

Reply:
If added a null check.